pkg/render/logstorage.go

```diff
@@ -1419,3 +1395,60 @@ func (es elasticsearchComponent) kibanaPodSecurityPolicy() *policyv1beta1.PodSec
 	psp.GetObjectMeta().SetName("tigera-kibana")
 	return psp
 }
+
+// If user has set ResourceRequirements, replace individual field's default value with user's value.
+// If user provided both Limits and Requests, use them.
+// As ResourceRequirements.Limits must be >= ResourceRequirements.Requests,
+// if user provided just Limits, and Limits is <= default Requests, set Requests value as user's Limits value,
+// if user provided just Requests, and Requests is >= default Limits, set Limits value as user's Requests value.
+func overrideResourceRequirements(defaultReq *corev1.ResourceRequirements, userOverrides corev1.ResourceRequirements) {
+	if _, ok := userOverrides.Limits["cpu"]; ok {
+		defaultReq.Limits["cpu"] = *userOverrides.Limits.Cpu()
+		if _, ok := userOverrides.Requests["cpu"]; !ok && defaultReq.Requests.Cpu().Value() > userOverrides.Limits.Cpu().Value() {
+			defaultReq.Requests["cpu"] = *userOverrides.Limits.Cpu()
+		}
+	}
+	if _, ok := userOverrides.Limits["memory"]; ok {
+		defaultReq.Limits["memory"] = *userOverrides.Limits.Memory()
+		if _, ok := userOverrides.Requests["memory"]; !ok && defaultReq.Requests.Memory().Value() > userOverrides.Limits.Memory().Value() {
+			defaultReq.Requests["memory"] = *userOverrides.Limits.Memory()
+		}
+	}
+	if _, ok := userOverrides.Requests["cpu"]; ok {
+		defaultReq.Requests["cpu"] = *userOverrides.Requests.Cpu()
+		if _, ok := userOverrides.Limits["cpu"]; !ok && defaultReq.Limits.Cpu().Value() < userOverrides.Requests.Cpu().Value() {
+			defaultReq.Limits["cpu"] = *userOverrides.Requests.Cpu()
+		}
+	}
+	if _, ok := userOverrides.Requests["memory"]; ok {
+		defaultReq.Requests["memory"] = *userOverrides.Requests.Memory()
+		if _, ok := userOverrides.Limits["memory"]; !ok && defaultReq.Limits.Memory().Value() < userOverrides.Requests.Memory().Value() {
+			defaultReq.Limits["memory"] = *userOverrides.Requests.Memory()
+		}
+	}
+}
+
+// Storage is the only valid resource requirement for PVC, extract that if user has set it.
+// If user provided both Limits and Requests, use them
+// As ResourceRequirements.Limits must be >= ResourceRequirements.Requests,
+// If user has provided just Limits, and Limits is <= default Requests, set Requests value as user's Limits value.
+// We don not set default Limits for storage, so don't have to handle case where user has set only Requests.
+func overridePvcRequirements(defaultReq *corev1.ResourceRequirements, userOverrides corev1.ResourceRequirements) {
```

Comment:
Similar to the override function above, I don't think we should be modifying the object through a parameter pointer, we should return something the resource requirements.